feat: Assign Event Tracking (M2-7607)

sultanofcardio opened the pull request:
- [x] Tests for the changes have been added

### 📝 Description

🔗 [Jira Ticket M2-7607](https://mindlogger.atlassian.net/browse/M2-7607)

This PR adds the following events to Mixpanel tracking:
- `Start Assign Activity/Flow`
- `Confirm Assign Activity/Flow`
- `Start Unassign Activity/Flow`
- `Confirm Unassign Activity/Flow`

It also updates the `Applet Save click` event to add some extra properties related to activity and flow assignments

### 📸 Screenshots

N/A

### 🪤 Peer Testing

Perform the following actions and confirm in the [Mindlogger Analytics Dev](https://mixpanel.com/project/3333580/view/3841316/app/events) project in Mixpanel that the proper events show up with the properties specified in the ticket:

- Perform an activity/flow assignment and un-assignment from the Applet Activities tab using the button in the toolbar
- Perform an activity/flow assignment and un-assignment from the Applet Activities tab using the more menu on the activity card
- Perform an activity/flow assignment and un-assignment from the Participant Assignment tab
- Perform an activity/flow assignment and un-assignment from the Data Viz page
- Make a minor change to the applet (like changing the name) and confirm that the `Applet Save click` event has the updated properties as specified in the ticket

### ✏️ Notes

- I’ve still implemented tracking assignment events on the Participant - Activities page for completeness, even though the two features are mutually exclusive
- This PR is a replacement for #1960 in order to get an Amplify environment running

farmerpaul committed:
feat: publish applet item type event tracking

farmerpaul committed:
refactor: add/reuse utility to track applet save

sultanofcardio committed:
Define new event types and add new props to the AppletSaveClickEvent type

sultanofcardio committed:
Track "Start Assign Activity/ Flow" clicks

sultanofcardio committed:
Track "Confirm Assign Activity/Flow" clicks

I've removed the `Entity Type` and `Activity ID`/`Activity Flow ID` properties for the ConfirmAssignActivityOrFlowEvent type because assignments can involve multiple activities and flows. Further, the event already has `Activity count` and `Flow count` properties

sultanofcardio committed:
Update tracked properties of the "Applet Save click" event

There was also an extra one of these events that gets triggered by the unsaved changes dialog. This is unnecessary, so I removed it.

sultanofcardio committed:
Track "Start Unassign Activity/Flow" clicks

sultanofcardio committed:
Track "Confirm Unassign Activity/Flow" clicks

sultanofcardio committed:
Update tests

sultanofcardio committed:
Merge branch 'develop' into feature/M2-7607-assign-event-tracking

farmerpaul commented on src/modules/Dashboard/components/ActivityAssignDrawer/ActivityAssignDrawer.tsx:
```diff
@@ -325,6 +326,37 @@ export const ActivityAssignDrawer = ({
+
+            if (assignment.activityFlowId) {
+              flowCount++;
+            }
```
The idea with tracking activity & flow counts is to get a sense of how many _unique_ activities/flows a user is assigning at one time – essentially, how many checkboxes were checked in the `ActivitiesList`. But this code will return a value of activity/flow count multiplied by the number of rows in the Assignments table. I saw how the wording in the Jira task wasn't 100% clear about that, so I went ahead and tweaked it to include the word "unique" for these properties.

Maybe all you need to do is just pass `activityIds.length` and `flowIds.length` below?

sultanofcardio replied:
That's a good call. Thanks for pointing this out, I overlooked that

farmerpaul commented on src/modules/Dashboard/features/RespondentData/RespondentDataHeader/RespondentDataHeader.tsx:
```diff
@@ -86,6 +92,13 @@ export const RespondentDataHeader = ({
+      action: MixpanelEventType.StartAssignActivityOrFlow,
+      [MixpanelProps.AppletId]: appletId,
+      [MixpanelProps.ActivityId]: activityId,
+      [MixpanelProps.EntityType]: 'activity',
```
This header component could be showing in the context of an activity or a flow. This would be one way to handle it:
```suggestion
      [activityFlowId ? MixpanelProps.ActivityFlowId : MixpanelProps.ActivityId]: activityFlowId || activityId,
      [MixpanelProps.EntityType]: activityFlowId ? 'flow' : 'activity',
```

farmerpaul reviewed (approved):
Great work. Code looks good, just noticed a couple issues with simple fixes, so preapproving!

sultanofcardio committed:
Merge branch 'develop' into feature/M2-7607-assign-event-tracking

sultanofcardio committed:
Make activity and flow count unique for the "Confirm Assign Activity/ Flow" event

sultanofcardio committed:
Add correct properties to "Start Assign Activity/ Flow" event based on whether it is an activity or flow

sultanofcardio committed:
Update failing tests

sultanofcardio committed:
Use `trackAppletSave` utility

sultanofcardio committed:
Fix type errors

sultanofcardio committed:
Update WithAppletSaveProps type

sultanofcardio committed:
Manually call onClickAssign

The click event was getting passed as the `activityOrFlow` parameter of the function

sultanofcardio committed:
Merge branch 'develop' into feature/M2-7607-assign-event-tracking

# Conflicts:
#	src/modules/Builder/features/SaveAndPublish/SaveAndPublish.hooks.test.tsx
#	src/modules/Builder/features/SaveAndPublish/SaveAndPublish.hooks.ts
#	src/modules/Dashboard/features/Participants/Participants.test.tsx
#	src/shared/utils/mixpanel/mixpanel.utils.ts

sultanofcardio committed:
Conditionally attach count properties

sultanofcardio committed:
Remove duplicate MixpanelFeature declaration

sultanofcardio committed:
Fix failing tests
